docs(world/modules/keysintable): initial version (#2959)
latticexyz/mud: docs/pages/world/modules/keysintable.mdx

import { CollapseCode } from "../../../components/CollapseCode";

# Keys in Table

[The `KeysInTable` module](https://github.com/latticexyz/mud/tree/main/packages/world-modules/src/modules/keysintable) tracks which keys are used with a specific table.
This is useful if you need to be able to access all the entries in a table onchain.
Note that it costs additional gas for every write to the table affected, so if you just need this functionality offchain it is best to look at the data offchain instead.

> ## Deployment

Author: No, we can't. Same problem as with the erc721 module, the module deployment happens before there is a table to manage.

```
(node:20571) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)

 Error while attempting deploy 

Error: Could not resolve dynamic resolution:
"0x746261707000000000000000000000005461736b730000000000000000000000"
    at k (file:///Users/qbzzt/run/keysInTable/node_modules/.pnpm/@latticexyz+world@2.0.12_@aws-sdk+client-kms@3.616.0_asn1.js@5.4.1_typescript@5.4.2/node_modules/@latticexyz/world/dist/internal.js:1:6856)
    at file:///Users/qbzzt/run/keysInTable/node_modules/.pnpm/@latticexyz+cli@2.0.12/node_modules/@latticexyz/cli/dist/commands-VSP7IDAM.js:16:142
    at Array.map (<anonymous>)
    at file:///Users/qbzzt/run/keysInTable/node_modules/.pnpm/@latticexyz+cli@2.0.12/node_modules/@latticexyz/cli/dist/commands-VSP7IDAM.js:16:135
    at async Promise.all (index 0)
    at async de (file:///Users/qbzzt/run/keysInTable/node_modules/.pnpm/@latticexyz+cli@2.0.12/node_modules/@latticexyz/cli/dist/commands-VSP7IDAM.js:15:2282)
    at async E (file:///Users/qbzzt/run/keysInTable/node_modules/.pnpm/@latticexyz+cli@2.0.12/node_modules/@latticexyz/cli/dist/commands-VSP7IDAM.js:20:132)
    at async file:///Users/qbzzt/run/keysInTable/node_modules/.pnpm/@latticexyz+cli@2.0.12/node_modules/@latticexyz/cli/dist/commands-VSP7IDAM.js:30:5370
```

> ## Deployment

Reviewer: > the module deployment happens before there is a table to manage

modules are installed as the last step after tables + systems are registered (but before post deploy script), so something else must be going on here

> ## Deployment

Author: I got it working.

import { Callout } from "nextra/components";

import { CollapseCode } from "../../../components/CollapseCode";

> # Keys in Table Module

Reviewer: ```suggestion
# Keys in Table module
```

[The `KeysInTable` module](https://github.com/latticexyz/mud/tree/main/packages/world-modules/src/modules/keysintable) tracks which keys are used with a specific table.
This is useful if you need to be able to iterate through all the records of a table onchain.

<Callout type="warning" emoji="💰">
It costs additional gas for every write to the table affected.
If you just need this functionality offchain it is best to look at the data offchain instead.

Also, this is a _generic_ solution.

> You can usually save gas by implementing a solution that is specific to your use case, using [hooks](/store/store-hooks).

Reviewer: its the hooks that are the costly pattern, its better to implement this manually and use-case-specific to save gas

Reviewer: can we remove the suggestion for hooks since that defeats the purpose of the custom solution

</Callout>

## Deployment

It can be deployed multiple times, each time for a different table.
For example, here is a `mud.config.ts` that deploys it for the `Tasks` table in [the React template](https://github.com/latticexyz/mud/tree/main/templates/react).

<CollapseCode>

```typescript filename="mud.config.ts" copy showLineNumbers {2,17-32}
import { defineWorld } from "@latticexyz/world";
import { resourceToHex } from "@latticexyz/common";

export default defineWorld({
  namespace: "app",
  tables: {
    Tasks: {
      schema: {
        id: "bytes32",
        createdAt: "uint256",
        completedAt: "uint256",
        description: "string",
      },
      key: ["id"],
    },
  },
  modules: [
    {
      root: true,
      artifactPath: "@latticexyz/world-modules/out/KeysInTableModule.sol/KeysInTableModule.json",
      args: [
        {
          type: "bytes32",
          value: resourceToHex({
            type: "table",
            namespace: "app",
            name: "Tasks",
          }), // end of value
        }, // end of arg
      ], // end of args
    }, // end of module

>   ], // end of modules

Reviewer: these comments are unnecessary noise, can we remove?

if we need to clarify this, we can:
- lift these out into functions/variables
- add syntax highlighting/hovers for brackets/nesting (like VSCode)

Author: > add syntax highlighting/hovers for brackets/nesting (like VSCode)

Do you know how to do that? That will be a good addition.

});
```

</CollapseCode>

<details>

<summary>Explanation</summary>

```typescript
import { resourceToHex } from "@latticexyz/common";
```

We need to provide `KeysInTable` with the resource ID for the table.

```typescript
  modules: [
    {
```

To install modules using `mud.config.ts`, you specify `modules:` which contains a list of module defintions.

```typescript
      root: true,
      artifactPath: "@latticexyz/world-modules/out/KeysInTableModule.sol/KeysInTableModule.json",
      args: [
```

- `artifactPath`, a link to the compiled JSON file for the module.
- `root`, whether to install the module with [root namespace permissions](/world/systems#root-systems) or not.
- `args` the module arguments.

```typescript
        {
          type: "bytes32",
          value:
```

The `args` list contains structures.
Each structure has two fields:

- `type`, the [Solidity data type](https://docs.soliditylang.org/en/latest/types.html)
- `value`, the value

```typescript

> resourceToHex({

Reviewer: why do we maintain indentation in code blocks above but not here? I think we should pick one or the other

Author: Oops, I meant to keep it.

  type: "table",
  namespace: "app",
  name: "Tasks",
}); // end of value
```

[This function](https://github.com/latticexyz/mud/blob/main/packages/common/src/resourceToHex.ts) creates the [resourceID](/world/resource-ids) for us from the table's data.

```typescript
        }     // end of arg
      ]       // end of args
    }         // end of module
  ]           // end of modules

> ```

Reviewer: not necessary to explain that we need to close each expression, can we remove?

      console.log(task.description);
    }
  }
}
```

</CollapseCode>

<details>

<summary>Explanation</summary>

```solidity
import { getKeysInTable } from "@latticexyz/world-modules/src/modules/keysintable/getKeysInTable.sol";
```

The function that gets the keys in the table.

```solidity
    StoreSwitch.setStoreAddress(worldAddress);
```

To use `getKeysInTable` we need to set the store address to `worldAddress`.

```solidity
    vm.startBroadcast(deployerPrivateKey);
    world.app__addTask("Walk the cat");
    world.app__addTask("Cook delicious meal");
    world.app__addTask("Practice Elvish");
    vm.stopBroadcast();
```

Creates three entries with different descriptions, to ensure we have entries to show.
Note that this is the _only_ place in the script we need to send transactions.
`getKeysInTable` is [a `view` function](https://docs.soliditylang.org/en/v0.8.26/contracts.html#view-functions), and does not change the state.

```solidity
    bytes32[][] memory keys = getKeysInTable(Tasks._tableId);
```

Get the key(s) by passing the Tasks table id as a parameter.

```solidity
    console.log("Number of keys:", keys.length);
    for(uint i=0; i<keys.length; i++) {

>        console.log("\nKey #", i, ":");

Reviewer: ```suggestion
       console.log("Key #", i, ":");
```
I think `console.log` already adds a new line?
(need to update the above script to match, too)

Author: `console.log` already adds *a* new line. But I want here to distinguish between different table rows more than I distinguish between different key fields of the same entry.

```
    for(uint i=0; i<keys.length; i++) {
       console.log("\nKey #", i, ":");
       for(uint j=0; j<keys[i].length; j++) {
          console.logBytes32(keys[i][j]);
       }
    }
```

Gives us 

```
Key #0:
0xdead
0xbeef

Key#1:
0x0bad
0x60a7
```

I think that is better than 

```
Key #0:
0xdead
0xbeef
Key#1:
0x0bad
0x60a7
```

       for(uint j=0; j<keys[i].length; j++) {
          console.logBytes32(keys[i][j]);
       }
    }
```

The `getKeysInTable` function returns a two-dimensional array.
The first dimension is the keys in the table.
The second dimension is the different key fields (in this case there is only one).

```solidity

>     console.log("\n\nTask descriptions");

Reviewer: I think its a bit easier to read if you separate the spacing from the meaningful log line
```suggestion
    console.log("\n\n");
    console.log("Task descriptions:");
```
(need to update the above script to match, too)

import { Callout } from "nextra/components";

import { CollapseCode } from "../../../components/CollapseCode";

# Keys in Table module

[The `KeysInTable` module](https://github.com/latticexyz/mud/tree/main/packages/world-modules/src/modules/keysintable) tracks which keys are used with a specific table.
This is useful if you need to be able to iterate through all the records of a table onchain.

> <Callout type="warning" emoji="💰">
> It costs additional gas for every write to the table affected.
> If you just need this functionality offchain it is best to look at the data offchain instead.
> 
> Also, this is a _generic_ solution.
> You can usually save gas by implementing a solution that uses your own `System` logic instead.
> 
> </Callout>

Reviewer: ```suggestion
<Callout type="warning" emoji="💰">
Using this module will add gas overhead to every write in the table it's used with.
You can reduce this overhead by implementing your own "onchain index" pattern.

</Callout>
```
